modules/bullet/collision_object_bullet.cpp

```diff
@@ -0,0 +1,265 @@
+/*************************************************************************/
+/*  collision_object_bullet.cpp                                          */
+/*  Author: AndreaCatania                                                */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                    http://www.godotengine.org                         */
+/*************************************************************************/
+/* Copyright (c) 2007-2017 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2017 Godot Engine contributors (cf. AUTHORS.md)    */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "collision_object_bullet.h"
+#include "area_bullet.h"
+#include "btBulletCollisionCommon.h"
+#include "bullet_physics_server.h"
+#include "bullet_types_converter.h"
+#include "bullet_utilities.h"
+#include "shape_bullet.h"
+
+#define enableDynamicAabbTree true
+#define initialChildCapacity 1
+
+CollisionObjectBullet::ShapeWrapper::~ShapeWrapper() {
+	bulletdelete(bt_shape)
```

Comment:
It looks like this broke compound shapes.  If I load a scene with a rigidbody and a single collision shape, add another shape to the rigidbody (thus creating a compound shape), and save the scene, I get the following error messages and a crash:
```
pure virtual method called
terminate called without an active exception
```

stack trace:
![image](https://user-images.githubusercontent.com/2694663/29744274-8f1db0d8-8a56-11e7-8e0e-f49ea4840e82.png)

It looks like saving the scene results in a queue_free() (no clue why),  which calls the ShapeWrapper destructor at some point (which frees the bullet shape), and eventually calls the virtual method shape->getMargin(), which fails because the bullet object was destroyed.

I can also reproduce this by calling queue_delete on a rigidbody with a compound shape when the game is running.

Removing "bulletdelete" fixes the problem, but that feels wrong as you do want to clean it up.  Maybe the fix would be to ensure shape->active is false after queue_free() is called?

Reply:
Just for log https://github.com/godotengine/godot/pull/10013#issuecomment-325632801

Reply:
That did the trick!